filebeat/input/file/state.go

```diff
@@ -30,14 +35,16 @@ func NewState(fileInfo os.FileInfo, path string, t string) State {
 		Timestamp:   time.Now(),
 		TTL:         -1, // By default, state does have an infinite ttl
 		Type:        t,
+		Meta:        meta,
 	}
 }
 
 // ID returns a unique id for the state as a string
 func (s *State) ID() string {
 	// Generate id on first request. This is needed as id is not set when converting back from json
 	if s.Id == "" {
-		s.Id = s.FileStateOS.String()
+		h, _ := hashstructure.Hash(s.Meta, nil)
+		s.Id = s.FileStateOS.String() + fmt.Sprintf("-%d", h)
```

Comment:
ouch... we optimized ID generation by removing Sprintf and reduce allocations. The ID generation used to be one of the worst offenders in CPU profiles. This change makes it worse again. How about:

```
if s.Meta == nil {
  s.Id = s.FileStateOS.String()
} else {
  hashValue, _ := hashstructure.Hash(s.Meta, nil)
  var hashBuf [17]byte
  hash, _ := strconv.AppendUint(h[:0], hashValue, 16)
  hash = append(hash, '-')

  fileID := s.FileStateOS.String()  

  var b strings.Builder
  b.Grow(len(hash) + len(fileID))
  b.Write(hash)
  b.WriteString(fileID)

  s.Id = b.String()
}
```

This still does execute an extra alloc, but only in the docker case and should be less expensive.

Reply:
sounds like a good idea, pushing this change